### PR TITLE
feat: asset slice upsert

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

**draft - wip**

- feat: wip makeAsset and upsertAsset
- test: better test

todo

- [x] requires https://github.com/shapeshift/lib/pull/1224 to expose `getRenderedIdenticonBase64`

currently, the asset service loads all assets from a statically generated file

as we pull in more data services, e.g. zerion/zapper/NFTs, we can't possibly have all supported assets at compile time, and need to upsert assets at runtime.

this is a POC with a utility function to generate assets from minimal data, such that they can be upserted.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

n/a

## Risk

reasonable - messes with asset service

### Engineering

* open for scrutiny

### Operations

* not testable by operations

## Screenshots (if applicable)

n/a

### CODE REVIEW

File: src/state/slices/assetsSlice/assetsSlice.ts [hunk 1]
```diff
@@ -57,6 +109,11 @@ export const assets = createSlice({
       state.byId = { ...state.byId, ...action.payload.byId } // upsert
       state.ids = Array.from(new Set([...state.ids, ...action.payload.ids]))
     },
+    upsertAsset: (state, action: PayloadAction<Asset>) => {
```
Comment (0xdef1cafe) [inline src/state/slices/assetsSlice/assetsSlice.ts:112]: consumers can now upsert assets into the store at runtime from external data sources, e.g. monkey pics/LP tokens

File: package.json [hunk 2]
```diff
@@ -90,7 +90,7 @@
     "@metamask/detect-provider": "^1.2.0",
     "@react-spring/web": "^9.5.2",
     "@reduxjs/toolkit": "^1.8.2",
-    "@shapeshiftoss/asset-service": "^8.10.0",
+    "@shapeshiftoss/asset-service": "^8.11.0",
```
Comment (0xdef1cafe) [inline package.json:93]: exposes `getRenderedIdenticonBase64`

File: src/state/slices/assetsSlice/assetsSlice.ts [hunk 3]
```diff
@@ -48,6 +51,55 @@ export const defaultAsset: Asset = {
...
+ * required fields are assetId, symbol, name, precision
+ * the rest can be inferred from existing data
+ */
+export const makeAsset = (minimalAsset: MinimalAsset): Asset => {
+  const { assetId } = minimalAsset
+
+  const color = (() => {
+    if (minimalAsset.color) return minimalAsset.color
+    const shaAssetId = sha256(assetId)
+    return `#${shaAssetId.slice(0, 6)}`
```
Comment (0xdef1cafe) [inline src/state/slices/assetsSlice/assetsSlice.ts:68]: tapshead.gif

**Review by gomesalexandre — APPROVED** [review]
Looks sane to me and unconsumed yet - paranoido-sanity checked that assets were looking good after a persisted store purge, LGTM

**Pull request merged**